### DISCUSSION TITLE
[Proposal] Keyword for inline IL code

### DISCUSSION THREAD

**Discussion by bzglinicki** [dotnet/csharplang#5723] [Language Ideas]
I want to suggest a new keyword `il` giving the ability to embed IL instructions within the source code written in C#, in the spirit of the `asm` / `__asm` keyword in C/C++ (see https://docs.microsoft.com/en-us/cpp/assembler/inline/inline-assembler).

The `il` keyword could appear wherever a C# statement is legal. It would have to be followed by an IL instruction or a group of IL instructions enclosed in braces (or an empty pair of braces):

```
il IL-instruction

il {
   Here goes some IL code.
}
```

Inline IL code could be integrated with C# code, it could use any C# variable, method name etc. that is available in the current scope.

I think this would be a very useful, elegant and easy-to-implement feature, complementary to the `System.Reflection.Emit` tools. I also believe that there is a need for such a functionality (see eg. https://stackoverflow.com/questions/3328901/how-to-insert-cil-code-to-c-sharp or https://stackoverflow.com/questions/26312026/execute-net-il-code-in-c-sharp).

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/discussions/226 https://github.com/dotnet/csharplang/discussions/927 https://github.com/dotnet/csharplang/discussions/4395

**Reply by jmarolf** [reply]
I think @tannergooding summed up my thoughts best with https://github.com/dotnet/csharplang/discussions/4395#discussioncomment-366279. @bzglinicki what can you not do today that inline IL would help you with?

**Reply by bzglinicki** [reply]
We can live without it, but I think inline IL would help make the source code more readable in some cases, e.g. replacing statements that use `System.Reflection.Emit` classes (when it's possible). As @canton7 mentioned in https://github.com/dotnet/csharplang/discussions/4395#discussioncomment-347584, there were 50,000 downloads of InlineIL.Fody by Feb 7, 2021, now there are over 96,000, so some people find inline IL useful.

**Reply by tannergooding** [reply]
There are certainly some cases that would be more readable; but that also is an increasingly minor number of cases; particularly as the language has brought on support for things that it couldn't express (like function pointers) and the BCL has exposed APIs (`System.Runtime.CompilerServices.Unsafe`) for others.

Its a lot of effort/functionality for something where things like `Fody` already exist to cover the few scenarios that remain.

And as I called out in 4395; using things like InlineIL can be very undesirable. Much of the time such patterns aren't well tested or well optimized and what C# is already doing/emitting is often better.
